### ISSUE TITLE
node-red crashes with dashboard

### ISSUE THREAD

**Issue by Karaburgut** [node-red/node-red#1125]
Hello. After started "node-red-start" I see WEB interface on localhost:1880 and dashboard on localhost:1880-ui. But after aprox. 2 minutes node-red go down and restart. And again, again.. with this log:

PS: I using mosquito MQTT brooker.

```
Started Node-RED graphical event wiring tool..
21 Jan 10:47:46 - [info]
Welcome to Node-RED
===================
21 Jan 10:47:47 - [info] Node-RED version: v0.16.1
21 Jan 10:47:47 - [info] Node.js  version: v6.9.4
21 Jan 10:47:47 - [info] Linux 4.4.38-v7+ arm LE
21 Jan 10:47:48 - [info] Loading palette nodes
root : TTY=unknown ; PWD=/root ; USER=root ; COMMAND=/usr/bin/python -u /usr/lib/node_modules/node-red/nodes/core/hardware/nrgpio.py info
pam_unix(sudo:session): session opened for user root by (uid=0)
pam_unix(sudo:session): session closed for user root
21 Jan 10:47:51 - [info] Dashboard version 2.2.1 started at /ui
21 Jan 10:47:53 - [info] Settings file  : /root/.node-red/settings.js
21 Jan 10:47:53 - [info] User directory : /root/.node-red
21 Jan 10:47:53 - [info] Flows file     : /root/.node-red/flows_raspberrypi.json
21 Jan 10:47:53 - [info] Server now running at http://127.0.0.1:1880/
21 Jan 10:47:53 - [info] Starting flows
21 Jan 10:47:53 - [info] Started flows
21 Jan 10:47:53 - [info] [mqtt-broker:719f17ae.438d78] Connected to broker: mqtt://localhost:1883
21 Jan 10:49:01 - [warn] Communication send error: Error: not opened
<--- Last few GCs --->
116585 ms: Mark-sweep 113.6 (144.5) -> 113.6 (144.5) MB, 1776.6 / 0.7 ms [allocation failure] [GC in old space requested].
118341 ms: Mark-sweep 113.6 (144.5) -> 113.5 (144.5) MB, 1755.7 / 0.6 ms [allocation failure] [GC in old space requested].
120174 ms: Mark-sweep 113.5 (144.5) -> 121.4 (136.5) MB, 1833.0 / 0.6 ms [last resort gc].
122842 ms: Mark-sweep 121.4 (136.5) -> 129.3 (136.5) MB, 2667.9 / 0.7 ms [last resort gc].
<--- JS stacktrace --->
==== JS stack trace =========================================
Security context: 0x22b7c315 <JS Object>
1: new constructor(aka Options) [/usr/lib/node_modules/node-red/node_modules/options/lib/options.js:~8] [pc=0x2967d4e4] (this=0x35808df9 <an Options with map 0x26efa8e9>,defaults=0x35808db1 <an Object with map 0x26efe595>)
3: initAsClient [/usr/lib/node_modules/node-red/node_modules/ws/lib/WebSocket.js:~571] [pc=0x29fa47f0] (this=0x35808cf9 <a WebSocket with map 0x5d510769>,address=0x3d8ea3...
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
nodered.service: main process exited, code=killed, status=6/ABRT
Unit nodered.service entered failed state.
nodered.service holdoff time over, scheduling restart.
Stopping Node-RED graphical event wiring tool....
Starting Node-RED graphical event wiring tool....
Started Node-RED graphical event wiring tool..
21 Jan 10:49:50 - [info]
```
Can you help me, someone?

Thnx

**Comment by knolleary** [comment]
Looks like an out of memory error.

Have you got a websocket node in your flow?

We've just fixed a memory leak in the ws node when it's set to connect to a remote websocket server - will be releasing 0.16.2 in the next couple days.

**Comment by Karaburgut** [comment]
Yes, I have two ws nodes. One for input and seconf for output. Can I configure this nodes correctly?

**Comment by Karaburgut** [comment]
That WS sending messages from UniPi [http://unipi.technology/wiki/Main_Page] via EVOK API [https://github.com/UniPiTechnology/evok].

**Comment by Karaburgut** [comment]
[https://github.com/UniPiTechnology/evok](url)

**Comment by Karaburgut** [comment]
[](url)

**Comment by knolleary** [comment]
The issue occurs when the ws node fails to connect to the remote system. There isn't a workaround other than to ensure the remote system is available for connections before Node-RED attempts to connect.

Will get 0.16.2 released today.

**Comment by Karaburgut** [comment]
Strange... I removed all WS nodes, but issue is still here...

**Comment by knolleary** [comment]
With exactly the same stack trace in the log? Please paste in here the lines after JS Stack Trace.

**Comment by Karaburgut** [comment]
`<--- JS stacktrace --->
==== JS stack trace =========================================
Security context: 0x5577c315 <JS Object>
1: _extend [util.js:~972] [pc=0x2bb554f4] (this=0x23a6a825 <an Object with map 0x3f027ad5>,target=0x59e3a39d <an Object with map 0x43c09b4d>,source=0x59e3a21d <an Object with map 0x3f0ea2cd>)
2: addRequest [_http_agent.js:123] [pc=0x3b4b70e8] (this=0x23a6ab75 <an Agent with map 0x3f013311>,req=0x71dd8479 <a ClientRequest with map 0x43c14911>,options=0x59e3a21d <an Object with map 0x3f0ea2cd...
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
nodered.service: main process exited, code=killed, status=6/ABRT
Unit nodered.service entered failed state.
nodered.service holdoff time over, scheduling restart.
Stopping Node-RED graphical event wiring tool....
Starting Node-RED graphical event wiring tool....
Started Node-RED graphical event wiring tool..
21 Jan 11:33:17 - [info]
`

**Comment by Karaburgut** [comment]
OK, I tried delete whole Flows. And start a new one. With MQTT and dashboard gauges only. That seems no issues, for now...

**Comment by knolleary** [comment]
We've now released 0.16.2 that contains the fix for the websocket memory leak. Given I believe that is what you hit, I'm going to close this issue. Please reopen if you hit it again after upgrading to 0.16.2.
